main/src/ca/uwaterloo/flix/language/ast/Ast.scala

```diff
@@ -366,55 +366,63 @@ object Ast {
   }
 
   /**
-    * Contains the edges held in a single constraint.
+    * Represents a labelled edge in the labelled graph.
+    * The labels represent predicate nodes that must co-occur for the dependency to be relevant.
     */
-  case class MultiEdge(head: (Name.Pred, Type), positives: Vector[TypedPredicate], negatives: Vector[TypedPredicate])
+  case class LabelledEdge(head: PredicateNode, p: Polarity, labels: Vector[PredicateNode], body: PredicateNode, loc: SourceLocation)
```

Comment:
Making head and body use PredicateNode is problematic because it includes the type which is not relevant.

Reply:
(also equality on types is not really defined because of type level Booleans.)

Reply:
Any reason why head and body cannot simply be predicate syms?

Reply:
It is to allow for easy switching between ignoring types, using arity, using full types etc. The types are never used directly, but uses the typeKeyOf function to decide equality, which right now returns the arity. But it could use that earlier to avoid keep track of superflous information. I will try that

Reply:
> Any reason why head and body cannot simply be predicate syms?

They need the arity information too otherwise we can only use more complicated filtering on the labels

Reply:
I am not sure I follow. In the paper they are just symbols. The labels are exclusively used for filtering. If the labels include arity then that information can be used.

Reply:
OK. I think I get it now. Let me think.

Reply:
While I understand "how it was", I don't see why we cannot refactor it to not include the type or whatever in the head/body. In one Datalog program we cannot have predicate symbols that disagree on their types.

Reply:
So while I understand how the previous version worked, I think it should be possible to align this with the paper, where we have LabelledEdges from symbols to symbols that carry extra information which tell us when they should be included.